### PR TITLE
Reporter options to truncate message and suppress stacktrace

### PR DESCRIPTION
**Pull request by james-stocks**
Signed-off-by: James Stocks <jstocks@chef.io>

### CODE REVIEW

File: lib/inspec/runner.rb
Comment (james-stocks) [inline lib/inspec/runner.rb]: Optionally the `run_data` could be directly modified here instead of passing options to the reporters. But I think there should be a distinction between the data that actually happened and the data that's reported

File: lib/inspec/reporters.rb
Comment (Schwad) [inline lib/inspec/reporters.rb]: I think what's interesting here, is wouldn't this setup give us the power to use options on any of our reporters in the future? (With the first two being these specific ones relating to JSON)

Though I do see where we could be increasing data coupling (passing data to the reporters) -  it is with a single options hash that opens up these future feature possibilities.

Comment (james-stocks) [reply lib/inspec/reporters.rb]: At this point, the `reporter` passed to this method is `["json", {"stdout"=>true}]`
That might be an existing hash that we can and should use to pass options for the reporter

File: lib/inspec/runner.rb
Comment (james-stocks) [inline lib/inspec/runner.rb]: From chatting to @kekaichinose the default for Automate should maybe be to not truncate. On the principle that for compliance, all data should be reported. 
Then Automate users will need to opt in for the reduced report size.

**Comment by alexpop** [comment]
Tried running inspec from this branch but getting this error:

```bash
[06:20:24 ~/git/inspec-stocksy (js/json)]$ bundle exec inspec exec ~/git/mycompliance-profile/myprofile1 --reporter json-automate
wrong number of arguments (given 2, expected 1)
```

If I switch to the system inspec, same profile runs:
```bash
[06:20:56 ~/git/inspec-stocksy (js/json)]$ inspec exec ~/git/mycompliance-profile/myprofile1 --reporter json-automate
{"platform":{"name":"mac_os_x","release":"19.4.0"},"profiles":[ ...
```

**Comment by james-stocks** [comment]
@alexpop @Schwad I changed the default behaviour to not change the output at all by default, and added some tests.
The remaining bit I need to do is either rename the new options to refer to JSON, or implement the option in all other reporters.

**Comment by Schwad** [comment]
Sorry I just saw the ready for review marking - I'll have a nice fresh look at this Monday if that works? Really excited for this feature! :)

**Comment by alexpop** [comment]
Thanks @james-stocks 

I tested the change and I can see it working if I run inspec like this:
```
[05:59:54 ~/git/inspec-stocksy (js/json)]$ bundle exec inspec exec ~/git/mycompliance-profile/myprofile1 --reporter_message_truncation 100 --reporter_backtrace_inclusion false --reporter json-automate
```

But passing the options via json-config does not truncate:
```
[05:59:54 ~/git/inspec-stocksy (js/json)]$ echo '{"reporter_backtrace_inclusion": false, "reporter_message_truncation": 100, "reporter":["json-automate"], "backend_cache":true, "chef_license": "accept"}' | bundle exec inspec exec ~/git/mycompliance-profile/myprofile1 --json-config=-
```

File: lib/inspec/runner.rb
Comment (clintoncwolfe) [inline lib/inspec/runner.rb]: This seems odd on first reading - check for a Symbol key, then set based on a String key?

Comment (clintoncwolfe) [inline lib/inspec/runner.rb]: These options are global to the run, not per-reporter. Inspec::Config (that is, `@conf`) is global, too - no need to make changes to `runner.rb` at all. You could just call it directly in `reporters/base.rb`

File: lib/inspec/reporters/base.rb
Comment (clintoncwolfe) [inline lib/inspec/reporters/base.rb]: Right here, you could call `runtime_config = Inspec::Config.cached` and get an instance of  the runtime config with access to the CLI options or JSON config - no need to sneak it into the per-reporter configs.

File: lib/inspec/reporters/base.rb
Comment (clintoncwolfe) [inline lib/inspec/reporters/base.rb]: Consider making this an error - if they provided the option but it is unusable, halt and show usage of the bad option is a reasonable thing to do. Continuing and generating too much data might be worse.

File: lib/inspec/reporters/base.rb
Comment (clintoncwolfe) [inline lib/inspec/reporters/base.rb]: I know naming things is hard and no one will ever agree, but perhaps "apply_message_size_limits"? If there are future options to implement, they would go in their own methods.

**Review by clintoncwolfe — COMMENTED** [review]
The algorithm seems fine, main thing I note is there is no need stash the CLI options on the per-reporter configs. Few other suggestions, most minor.

File: lib/inspec/reporters/base.rb
Comment (james-stocks) [reply lib/inspec/reporters/base.rb]: I was foreseeing the opposite - that this method _would_ handle any future options. I can rename the method like you suggest so that it is named specific to the thing it _currently_ does, and we can discuss handling future options when we get to that point.

File: lib/inspec/reporters/base.rb
Comment (james-stocks) [reply lib/inspec/reporters/base.rb]: Thanks, that sounds good! I think that solves the problem Alex mentioned where these options are not working from STDIN config.
It would make it a little odd that the class is instantiated with a `config` hash but we don't use that for config - but it does mean I can revert my changes to the runner.

File: lib/inspec/reporters/base.rb
Comment (james-stocks) [reply lib/inspec/reporters/base.rb]: My thinking for making this a warning is that:
 * The CLI, not this class, is responsible for config validation (even though it's not doing any currently...). 
 * At this point, the run has completed. It possibly took a long time and burned through billable compute resource. The reporter should generally never halt.

Possibly I could make the option strictly numeric and throw away the `ALL` special case - I did it this way so that those using InSpec in CI could have a variable driven run command where `message-truncation` is always set at the CLI and the value is defined by a CI variable.

File: lib/inspec/runner.rb
Comment (james-stocks) [reply lib/inspec/runner.rb]: I pushed another commit that leaves runner.rb unchanged

File: lib/inspec/reporters/base.rb
Comment (clintoncwolfe) [reply lib/inspec/reporters/base.rb]: Good point about the validation happening post-run - perhaps the validation should be moved to be here - https://github.com/inspec/inspec/blob/2e887a94afcca819da781d4774aa2a5a0b56785e/lib/inspec/config.rb#L328

Comment (james-stocks) [reply lib/inspec/reporters/base.rb]: Thanks, that looks like the right spot for validation.
This warning was meant to supplementary to proper validation - but I think if there's validation in `config.rb` then this class also doing a check is overkill - so I'll just remove this warning.

**Review by clintoncwolfe — APPROVED** [review]
This looks good, thanks!

File: lib/inspec/reporters/base.rb
Comment (Schwad) [inline lib/inspec/reporters/base.rb:14]: This is good to know for future

File: lib/inspec/reporters/base.rb
Comment (Schwad) [inline lib/inspec/reporters/base.rb:20]: I had a quick look with `benchmark/ips` in the Terminal with an identical data structure to see if I could find another way to make this change without the loops but as far as I could tell this is the best approach!

**Review by Schwad — APPROVED** [review]
Looking great, thanks! 👍

File: lib/inspec/config.rb
Comment (james-stocks) [inline lib/inspec/config.rb]: This is wrong if the value is already an Integer. Usually it will be a string (via the CLI) but can be an Integer via JSON config.
@clintoncwolfe 's PR for typing inputs will also cause this to be wrong, as that PR causes inputs to be typed where possible

**Review by Schwad — CHANGES_REQUESTED** [review]
Look to handle actual integer via JSON:

````
echo '{"reporter_backtrace_inclusion": false, "reporter_message_truncation": 100, "reporter":["json-automate"]}' | be inspec exec ~/git/mycompliance-profile/myprofile1 --json-config=-
#=> reporter_message_truncation is set to 100. It must be set to an integer value or ALL to indicate no truncation.
```

File: lib/inspec/reporters/base.rb
Comment (james-stocks) [reply lib/inspec/reporters/base.rb:20]: Yes I don't much like these nested loops.
Note the cautious  `&.` because at each level the report might be missing any occurrences of the nested elements.

**Comment by james-stocks** [comment]
Please don't merge this before @alexpop approves

File: lib/inspec/config.rb
Comment (clintoncwolfe) [reply lib/inspec/config.rb]: #5011 will only impact the `--input` CLI option

**Comment by alexpop** [comment]
Thank you everybody, great team effort!
```
$ echo '{"reporter_backtrace_inclusion": false, "reporter_message_truncation": 100, "reporter":["json-automate"]}' | bundle exec inspec exec ~/git/mycompliance-profile/myprofile1 --json-config=- | jq .
...
          "waiver_data": {},
          "results": [
            {
              "status": "failed",
              "code_desc": "Command: `ps aux` stdout is expected to include \"No packages needed for security\"",
              "run_time": 0.70555,
              "start_time": "2020-05-06T22:07:55+01:00",
              "message": "expected \"USER               PID  %CPU %MEM      VSZ    RSS   TT  STAT STARTED      TIME COMMAND\\nap[Truncated to 100 characters]"
            }
          ]
        }
      ],
      "status": "loaded"
    }
  ],
  "statistics": {
    "duration": 0.711616
  },
  "version": "4.18.109"
}
```

Same result with:
```
bundle exec inspec exec ~/git/mycompliance-profile/myprofile1 --reporter_message_truncation 100 --reporter json-automate | jq .
```

**Review by alexpop — APPROVED** [review]
Cheers 🚀 🚀 🚀

**Review by Schwad — APPROVED** [review]

File: lib/inspec/config.rb
Comment (james-stocks) [reply lib/inspec/config.rb]: D'oh, good point!

**Pull request merged**
